### ISSUE TITLE
Support for authenticating non-GitHub remotes

### ISSUE THREAD

**Issue by shiftkey** [desktop/desktop#852]
This is an open discussion what we should target for the beta, if anything. We can spin off specific tasks and put those into the appropriate milestones.

Some things that should cover most of the use cases we've seen based on Desktop Classic:

 - let the user specify credentials when an authentication fails?
 - managing credentials - list stored creds, let user remove

Aside from that, if we're keen to do this we'd need to update our `getUserForRepository` lookup to be more flexible (e.g. if not marked as a GitHub repository, lookup for credentials by the remote URL)

**Comment by blogcraft** [comment]
@joshaber When you go to [Github Desktop](https://desktop.github.com/) The first thing that appears is the Beta, and hidden in the bottom is the current one. So, yeah, it might be a beta, but it is promoted as a first choice. So non-GitHub support is to be expected.

**Comment by bjcull** [comment]
I use both Github and VSTS repos with classic quite often. I was a little disappointed to find I can't use this for VSTS. Is a stop-gap of calling the underlying git commands feasible? git push works just fine from the shell.

**Comment by tech4him1** [comment]
I don't think this is necessary right away, but I still think it is very important if GitHub Desktop is going to really be used as more than a hobby tool.

**Comment by fengsong97** [comment]
今天更新了 deskTop客户端，一直挺喜欢这个客户端的， 但是这次更新发现，如果不是 github 仓库的话 ， 就无法pull 和 push ，需要认证，但是却没有提供认证界面，
 这真的让人失望 ，希望开发人员能够快点解决这个问题。

Today updated the deskTop client, the client has been very love, but if not found this update, GitHub warehouse, to pull and push, need to be Authentication, but did not provide authentication interface,
It's really disappointing, and I hope developers can work it out quickly.

![image](https://user-images.githubusercontent.com/12808002/27126808-3ceed85a-512b-11e7-8628-4c8d1a66542a.png)

**Comment by mikeytusa** [comment]
Definitely need to be able to use this with non-GitHub repos.

**Comment by FelschR** [comment]
I really like GitHub Desktop but I also have to work with VSTS, Bitbucket & GitLab at work for different projects and it would be really nice if GitHub Desktop would support that as well.

**Comment by fengsong97** [comment]
The GithubDeskTop of old version link for non-GitHub can be supported . 
example: oschina.com  gitlab .....

链接:http://pan.baidu.com/s/1jH4l3Lk  密码:v9ag

**Comment by 47ronin** [comment]
Confirmed that the beta cannot properly authenticate against Pantheon Drush repos in order to clone dev projects. It doesn't error out. It just throws a never-ending spinner icon in the dialog.

**Comment by JimMorrison723** [comment]
Is there any ETA for this (non-Github repo support)? Or should i just go back to the 'old' client and wait for the first stable version? :)

**Comment by iAmWillShepherd** [comment]
@JimMorrison723 sorry for the long delay. Most of the team has been out on vacation, so that will affect the date of completion. Unfortunately, I cannot give you an ETA, but we will be addressing this in the [Milestone for 1.0](https://github.com/desktop/desktop/milestone/7).

**Comment by EliasCK** [comment]
Are there any work around's to get non-Github repo's working. I have a private bitbucket repo I would like to sync up.

**Comment by trevorhreed** [comment]
@EliasCK While the beta doesn't support this, you can use the current stable version to connect to private BitBucket repos. Look at the bottom of this page: https://desktop.github.com/

**Comment by alanaasmaa** [comment]
I have also couple of repos in bitbucket. It would be awesoe to atleast have some kind of support for it.

**Comment by mrahhal** [comment]
For people who need this to work with vsts, you need to got to "Security" and then to "Alternate Credentials" and enable it (make sure to use a different password).

**Comment by mikeytusa** [comment]
Is this supposed to be working now as of 0.8.0?  I see it in the release notes but it's definitely not working for me on Windows.

**Comment by shiftkey** [comment]
@mikeytusa please open a new issue with details about your setup

**Comment by FelschR** [comment]
@mikeytusa It's working for me.
It only work with HTTPS and not SSH (at the moment).

**Comment by mikeytusa** [comment]
@FelschR ah, yes that would be it.  I'm using SSH for authentication.

**Comment by nagman** [comment]
It doesn't work for me neither. I can't connect to my private git repo.
Windows -> Github Desktop -> ssh to my private git repo on a local NAS.

**Comment by Growiel** [comment]
Working for me in HTTPS, not at all in SSH, as already mentioned.

**Comment by 47ronin** [comment]
Works for Pantheon's Drush-based repositories but requires a few extra steps:

1. On Mac, you may get a brief, non-interactive error when the repo asks for login: Error: `ssh_askpass: exec(/usr/X11R6/bin/ssh-askpass): No such file or directory`
2. Resolve this by first installing https://github.com/theseal/ssh-askpass
3. Even after following the install instructions, you may still get the error
4. Create a symlink (if you installed ssh-askpass via Homebrew):
```
$ cd /usr/X11R6/bin
$ sudo ln -s /usr/local/bin/ssh-askpass ssh-askpass
```
This works, and creates a password modal in GitHub Desktop (beta). However, the password is never saved so it must be entered with each transaction.

**Comment by HedvigS** [comment]
Is there any news about GitHub Desktop being able to handle other remote services, like GitLab? What's the plan? 

I tried reading this thread, but it's quite hard to follow.

**Comment by mikeytusa** [comment]
On a PC it doesn't support SSH.  Works over HTTPS. It works on Mac though.  Not sure why it hasn't been implemented yet on a PC.

**Comment by HedvigS** [comment]
Thanks @mikeytusa . How does it work on a mac?

**Comment by shiftkey** [comment]
For those who are looking at SSH support, we're tracking the current pain points over here #2579.

If you're using HTTPS, the work done to support basic credentials from other providers has landed in 1.0.

**Comment by 47ronin** [comment]
Seeing a new `ssh-askpass` error when attempting to fetch/sync in the new GitHub desktop app?

If you are using macOS 10.13 High Sierra and `/usr/X11R6` disappeared, don't panic. Yes it's gone and symbolic linking in` /usr` cannot be done due to SIP. There is a simple workaround:
https://github.com/theseal/ssh-askpass/issues/18#issuecomment-335617082
